Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/preprocessor/CqlPreprocessorElmCommonVisitor.kt

```diff
@@ -0,0 +1,878 @@
+    }
+
+    override fun visitIntervalTypeSpecifier(
+        ctx: IntervalTypeSpecifierContext
+    ): IntervalTypeSpecifier {
+        val result =
+            of.createIntervalTypeSpecifier().withPointType(parseTypeSpecifier(ctx.typeSpecifier()))
+        val intervalType = IntervalType(result.pointType.resultType)
+        result.resultType = intervalType
+        return result
+    }
+
+    override fun visitListTypeSpecifier(ctx: ListTypeSpecifierContext): ListTypeSpecifier {
+        val result =
+            of.createListTypeSpecifier().withElementType(parseTypeSpecifier(ctx.typeSpecifier()))
+        val listType = ListType(result.elementType.resultType)
+        result.resultType = listType
+        return result
+    }
+
+    fun parseFunctionHeader(ctx: FunctionDefinitionContext): FunctionHeader {
+        val functionDef =
+            of.createFunctionDef()
+                .withAccessLevel(parseAccessModifier(ctx.accessModifier()))
+                .withName(parseString(ctx.identifierOrFunctionIdentifier()))
+                .withContext(currentContext)
+        if (ctx.fluentModifier() != null) {
+            libraryBuilder.checkCompatibilityLevel("Fluent functions", "1.5")
+            functionDef.isFluent = true
+        }
+        if (ctx.operandDefinition() != null) {
+            for (opdef in ctx.operandDefinition()) {
+                val typeSpecifier = parseTypeSpecifier(opdef.typeSpecifier())
+                functionDef.operand.add(
+                    of.createOperandDef()
+                        .withName(parseString(opdef.referentialIdentifier()))
+                        .withOperandTypeSpecifier(typeSpecifier)
+                        .withResultType(typeSpecifier!!.resultType) as OperandDef
+                )
+            }
+        }
+        val typeSpecifierContext = ctx.typeSpecifier()
+        return if (typeSpecifierContext != null) {
+            FunctionHeader.withReturnType(functionDef, parseTypeSpecifier(typeSpecifierContext))
+        } else FunctionHeader.noReturnType(functionDef)
+    }
+
+    protected fun parseTypeSpecifier(pt: ParseTree?): TypeSpecifier? {
+        return if (pt == null) null else visit(pt) as TypeSpecifier
+    }
+
+    protected fun parseAccessModifier(pt: ParseTree?): AccessModifier {
+        return if (pt == null) AccessModifier.PUBLIC else (visit(pt) as AccessModifier)
+    }
+
+    protected fun parseQualifiers(ctx: NamedTypeSpecifierContext): List<String?> {
+        val qualifiers: MutableList<String?> = ArrayList()
+        if (ctx.qualifier() != null) {
+            for (qualifierContext in ctx.qualifier()) {
+                val qualifier = parseString(qualifierContext)
+                qualifiers.add(qualifier)
+            }
+        }
+        return qualifiers
+    }
+
+    protected open fun getModel(
+        modelNamespace: NamespaceInfo?,
+        modelName: String?,
+        version: String?,
+        localIdentifier: String?
+    ): Model? {
```

Comment:
question: This smells like an incorrect inference. All the parameters are nullable